### DISCUSSION TITLE
About the stability issue of shape matching method

### DISCUSSION THREAD

**Discussion by fangzhouzisb** [sofa-framework/sofa#4088] [Develop in C++]
Hi there,
I noticed the [Shape Matching plugin](https://www.sofa-framework.org/applications/marketplace/shape-matching/), @fredroy, in Sofa, and I believe Shape Matching is a promising deformation algorithm. However, upon testing it, I've observed that it seems to be unstable (experiencing numerical explosions with even slight pull and drag deformations). Additionally, this plugin lacks the implementation of the addKToMatrix virtual function, so the FreeMotionAnimationLoop and SparseLDLSolver components cannot be in use. I attempted to implement the addKToMatrix function myself, but my efforts ended in failure (resulting in numerical explosions). I'm unsure if this is due to my own skill level or if Shape Matching itself is inherently unstable under the implicit Euler method. I would greatly appreciate it if someone could help me address this issue. 
Thank you.

**Comment by alxbilger** [comment]
Does your scene work with a `CGLinearSolver`? You can try to increase the number of iterations of the CG, and decrease the thresholds so that the computed solution is closer to the "right" solution. If you change the solver for a direct one, you should end up with very similar result. I added `ShapeMatchingForceField` to [this list](https://github.com/sofa-framework/sofa/issues/3967) so I should work on it during the next months.

@hugtalbot What is the status of this plugin? Is it still maintained?

**Comment by alxbilger** [comment]
Regarding your implementation of `addKToMatrix`, the maths are quite challenging because of the rotation matrix. Do you have any resource with the maths to derivate the expression of the force?
